### PR TITLE
Add Storage Units, Stored Property Insurance to the flow

### PR DESCRIPTION
**Pull request by vr8686**
1. fixed test failing issue
2. added environmental fee collection on shopping cart page

### CODE REVIEW

File: uhaul/components/order_models/truck_order.py [hunk 1]
```diff
@@ -5,11 +5,15 @@ class TruckOrder(Order):
     """Truck Order Model"""
 
     def __init__(self):
-        self.pick_up_location: str = '33101'
-        self.pick_up_city: str = 'Miami, FL'
-        self.drop_off_location: str = '33101'
+        self.pick_up_location: str = '08816'
+        self.pick_up_city: str = 'East Brunswick, NJ'
+        self.drop_off_location: str = '08816'
```
[the remark this answers is not in the document]
Comment (vr8686) [reply uhaul/components/order_models/truck_order.py:10]: Later I will add result round as @ehnat0n suggested. It should work for every single place, not just Miami ;)

File: uhaul/components/trucks/order_option.py [hunk 2]
```diff
@@ -1,10 +1,12 @@
+import time
+
 from selenium.webdriver.chrome.webdriver import WebDriver
 
 from base.components.base import Base
 
 
 class OrderOption(Base):
-    NO_THANKS = '//*[self::a or self::button][starts-with(normalize-space(.), "No thanks, I do not need")]'
+    NO_THANKS = '//*[self::a or self::button][starts-with(normalize-space(.), "No thanks,")]'
```
[the remark this answers is not in the document]
Comment (vr8686) [reply uhaul/components/trucks/order_option.py:9]: When I ran test behave could not find xpath with ```"No thanks, I do not need"```. I have no idea why..I shortened it and it works now.

File: uhaul/components/trucks/storage_units.py [hunk 3]
```diff
@@ -1,9 +1,82 @@
...
+
+        #  add insurance price to truck price record
+        self.collect_insurance_price()
+
+        #  click "Continue" button
+        self.find_element(self.CONTINUE_BUTTON).click()
+
+    def collect_price(self):
+        price = float(
+            self.find_element(self.STORAGE_PRICE.format(unit_size=self.order.unit_size)).text.replace('$', ''))
```
[the remark this answers is not in the document]
Comment (vr8686) [reply uhaul/components/trucks/storage_units.py:53]: I added this code to my ``àdd_option```method to my latest commit.
It's interesting, but looks like there are 2 options how to handle the data from feature file:
1) as you did - used it within your ``àdd_option``` method. You used ``òption```argument:
```
for key, value in options.items():
            select = Select(self.find_element(self.AMOUNT_OPTION.format(option=key)))
            select.select_by_visible_text([option.text for option in select.options if value in option.text][0])

            price_str = (self.find_element(self.PRICE.format(option=key)).text.replace('$', '')).split(
                ' ')[0]
            amount = Fraction(eval(value.split(' ')[0]))
            price = float(price_str) * float(amount)

            self.order.truck_price_records |= {key: price}

        self.click(self.BUTTON_ADD)
```

OR

2) we can add it to `class TruckOrder(Order):` as @arturadamian did here:
```
@step('Create a new {order_type:OrderType} order')
def create_new_order(context, order_type: Type[TruckOrder | StorageOrder]) -> None:
    BS(context).set_context_var('order', order_type())
    context.order.set_data(ContextTable(context).to_flat_dict())
```

idk which option is correct, honestly.. 
I added this code to my ``àdd_option```method and just update date in `class TruckOrder(Order):`

 ```
def add_option(self, options):
        for k, v in options.items():
            option = {k: v}
            self.order.set_data(option)

```

File: uhaul/components/trucks/moving_option.py [hunk 4]
```diff
@@ -1,9 +1,13 @@
...
 class MovingOption(OrderOption):
 
-    def __init__(self, driver: WebDriver):
-        super().__init__(driver)
\ No newline at end of file
+    def __init__(self, driver: WebDriver, order: TruckOrder = None):
+        super().__init__(driver)
+
+    def add_option(self):
+        pass
```
[the remark this answers is not in the document]
Comment (vr8686) [reply uhaul/components/trucks/moving_option.py:13]: by the way, I am going to use @dvlad7909's version witm match - case. it is more flexible in this particular case.

File: uhaul/components/trucks/storage_units.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (dvlad7909) [reply uhaul/components/trucks/storage_units.py:61]: Wow, thats cool!

File: uhaul/components/trucks/shopping_cart.py [hunk 5]
```diff
@@ -20,3 +22,8 @@ def verify_price_due_at_pickup(self):
         assert actual_price == expected_price, (f"Due at Pick Up price is not as expected:"
```
[the remark this answers is not in the document]
Comment (vr8686) [reply uhaul/components/trucks/shopping_cart.py:22]: Let’s ask @arturadamian

**Review by dvlad7909 — APPROVED** [review]

File: uhaul/components/trucks/shopping_cart.py [hunk 6]
```diff
@@ -9,6 +9,8 @@ class ShoppingCart(Table):
     DUE_TODAY = '//div[contains(@class, "divider")][.//dt[contains(., "Due Today")]]//dd'
     DUE_AT_PICKUP = ('//div[contains(@class, "secondary")]'
                      '//dt[contains(., "Equipment Rental")]/following-sibling::dd[position() = 1]')
+    ENVIRONMENTAL_FEE = ('//table[contains(@class, "cart")]'
+                         '//td[contains(., "Environmental")]/following-sibling::td[position() = 2]//b')
```
[the remark this answers is not in the document]
Comment (vr8686) [reply uhaul/components/trucks/shopping_cart.py:13]: I agree with you, `class='text-right'` is more reliable. 
I was just following the given code structure.

File: uhaul/components/trucks/storage_units.py [hunk 7]
```diff
@@ -1,9 +1,82 @@
...
+from dateutil import parser
 from selenium.webdriver.chrome.webdriver import WebDriver
+from selenium.webdriver.common.by import By
 
+from uhaul.components.order_models.truck_order import TruckOrder
 from uhaul.components.trucks.order_option import OrderOption
 
 
 class StorageUnitsOption(OrderOption):
+    STORAGES = '//div[@class="form-content"]//li[@class="cell"]'
```
[the remark this answers is not in the document]
Comment (vr8686) [reply uhaul/components/trucks/storage_units.py:16]: Agreed. More reliable too.

File: uhaul/components/trucks/storage_units.py [hunk 8]
```diff
@@ -1,9 +1,82 @@
...
+    A_STORAGE = STORAGES + '//h4[contains(text(), "{unit_size}")]'
+    STORAGE_PRICE = A_STORAGE + ('/parent::div[@class="cell auto"]/following-sibling::'
+                                 'div[contains(@class, "medium-shrink")]//b')
```
[the remark this answers is not in the document]
Comment (vr8686) [reply uhaul/components/trucks/storage_units.py:19]: Well, for me, `text-right` looks absolutely similar as `medium-shrink` - just some nodes with no specific name. I got `text-right` first, but then changed to `medium-shrink`. The way you describe it sounds logical. I'll take that. thank you.

File: uhaul/components/trucks/storage_units.py [hunk 9]
```diff
@@ -1,9 +1,82 @@
...
+    RESERVE_BUTTON = A_STORAGE + '/ancestor::div[@class="grid-x grid-full"]//label[contains(text(), "Rent Now")]'
```
[the remark this answers is not in the document]
Comment (vr8686) [reply uhaul/components/trucks/storage_units.py:20]: I'm glad you are reviewing it. I was so blind until you pointed it out :)

File: uhaul/components/trucks/storage_units.py [hunk 10]
```diff
@@ -1,9 +1,82 @@
...
+    CALENDAR_OPEN = A_STORAGE + ('/ancestor::div[@class="grid-x grid-full"]'
+                                 '//div[contains(@id, "rentNowContainer")]//input[@name="MoveInDate"]')
+    CALENDAR_MODAL = '//div[@id="ui-datepicker-div"]'
+    CALENDAR_PAGE = '//div[contains(@id, "ui-datepicker")]'
+    DATE = '//a[@class="ui-state-default"][@data-date="{day}"]'
+    ADD_BUTTON = A_STORAGE + ('/ancestor::div[@class="grid-x grid-full"]'
+                              '//div[contains(@id, "rentNowContainer")]//button[contains(text(), "Add to Cart")]')
+    CONTINUE_BUTTON = '//button[contains(text(), "Continue")]'
+    INSURANCE = '//ul[@class="storage-insurance-choice no-bullet"]//span[contains(text(), "{insurance}")]'
+    INSURANCE_PRICE = INSURANCE + '/ancestor::div[@class="cell medium-auto"]/following-sibling::div//strong'
```
[the remark this answers is not in the document]
Comment (vr8686) [reply uhaul/components/trucks/storage_units.py:30]: Screenshot 2023-12-13 at 8 58 42 PM
not in this case, unfortunately.

File: uhaul/components/trucks/storage_units.py [hunk 11]
```diff
@@ -1,9 +1,82 @@
...
+    STORAGE_PRICE = A_STORAGE + ('/parent::div[@class="cell auto"]/following-sibling::'
+                                 'div[contains(@class, "medium-shrink")]//b')
+    RESERVE_BUTTON = A_STORAGE + '/ancestor::div[@class="grid-x grid-full"]//label[contains(text(), "Rent Now")]'
+    CALENDAR_OPEN = A_STORAGE + ('/ancestor::div[@class="grid-x grid-full"]'
+                                 '//div[contains(@id, "rentNowContainer")]//input[@name="MoveInDate"]')
+    CALENDAR_MODAL = '//div[@id="ui-datepicker-div"]'
+    CALENDAR_PAGE = '//div[contains(@id, "ui-datepicker")]'
+    DATE = '//a[@class="ui-state-default"][@data-date="{day}"]'
+    ADD_BUTTON = A_STORAGE + ('/ancestor::div[@class="grid-x grid-full"]'
+                              '//div[contains(@id, "rentNowContainer")]//button[contains(text(), "Add to Cart")]')
```
[the remark this answers is not in the document]
Comment (vr8686) [reply uhaul/components/trucks/storage_units.py:27]: All the buttons are the same on the page, only the label differs. That's why I used it :)

**Review by ehnat0n — APPROVED** [review]
Good job!

File: requirements.txt [hunk 12]
```diff
@@ -4,7 +4,4 @@ webdriver-manager>=4.0.1
 urllib3==1.26.6
 requests==2.31.0
 jsonpath-ng>=1.6.0
-python-dateutil~=2.8.2
-reportportal-client==5.4.0
-behave-reportportal==3.0.0
```
Comment (In-Anna) [inline requirements.txt:9]: I guess you removed these during merge conflict resolving...
This is why you have 1 failed check - you can go to details and check it out
Screenshot 2023-12-17 at 11 45 11 PM

Comment (vr8686) [reply requirements.txt:9]: omg, thank you!

Comment (vr8686) [reply requirements.txt:9]: fixed that, thanks again!

**Review by In-Anna — APPROVED** [review]
lgtm

**Comment by vr8686** [comment]
@In-Anna shall I merge it?

**Comment by In-Anna** [comment]
> @In-Anna shall I merge it?

Sure, you have all approvals

**Pull request merged**